###############################################################################
# Copyright (c) 2008, 2013, Oracle and/or its affiliates. All rights reserved.
#
# Permission is hereby granted, free of charge, to any person obtaining a
# copy of this software and associated documentation files (the "Software"),
# to deal in the Software without restriction, including without limitation
# the rights to use, copy, modify, merge, publish, distribute, sublicense,
# and/or sell copies of the Software, and to permit persons to whom the
# Software is furnished to do so, subject to the following conditions:
#
# The above copyright notice and this permission notice (including the next
# paragraph) shall be included in all copies or substantial portions of the
# Software.
#
# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.  IN NO EVENT SHALL
# THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
# FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
# DEALINGS IN THE SOFTWARE.
#

5039226 Whitebox animation causes Solaris desktop hang

This problem is caused by a dead lock between java_vm and Xserver:
Xserver is grabbed by java_vm and waiting for XUngrabServer, java_vm
thought XUngrabServer was already sent and sends another dga lock
request to Xserver and blocked on that request. Here is what happened:
java_vm threads calls XGrabServer and XUngrabServer and then
Solaris_DGA_GetLock, it seems no problem, but the current
XUngrabServer requires a XSync call after XUngrabServer in MT
environment to make sure Xserver is ungrabbed *before* new lock
requests come in. Without this XSync, a dead lock may formed just like
this case.

diff --git a/src/UngrabSvr.c b/src/UngrabSvr.c
index ea00848..cea7f99 100644
--- a/src/UngrabSvr.c
+++ b/src/UngrabSvr.c
@@ -39,6 +39,9 @@ XUngrabServer (
         GetEmptyReq(UngrabServer, req);
         UnlockDisplay(dpy);
 	SyncHandle();
+	/* fix for bug 5039226: desktop hangs */
+	if (!dpy->synchandler)
+	    XSync(dpy, 0);
 	return 1;
 }
 
